From e3856360c69f4610e6b508ae700ce482f5808e89 Mon Sep 17 00:00:00 2001
From: Long Term Stable Dev <mitch.harder@sabayonlinux.org>
Date: Tue, 3 Jul 2012 13:19:40 -0500
Subject: [PATCH] Patch64: glibc-rh769594.patch

glibc-rh769594.patch
2011-11-14  Andreas Schwab  <schwab@redhat.com>

	* malloc/arena.c (arena_get2): Don't call reused_arena when
	_int_new_arena failed.

2011-11-10  Andreas Schwab  <schwab@redhat.com>

	* malloc/arena.c (_int_new_arena): Don't increment narenas.
	(reused_arena): Don't check arena limit.
	(arena_get2): Atomically check arena limit.

commit a5fb313cb7b7e692fd4684916aaa98e03ec7e8b6
Author: Andreas Schwab <schwab@redhat.com>
Date:   Mon Nov 14 11:41:52 2011 +0100

    Don't call reused_arena when _int_new_arena failed
---
 malloc/arena.c |   65 ++++++++++++++++++++++++++++++-------------------------
 1 files changed, 35 insertions(+), 30 deletions(-)

diff --git a/malloc/arena.c b/malloc/arena.c
index 4d0deef..ae5a143 100644
--- a/malloc/arena.c
+++ b/malloc/arena.c
@@ -947,8 +947,6 @@ _int_new_arena(size_t size)
   main_arena.next = a;
 
 #ifdef PER_THREAD
-  ++narenas;
-
   (void)mutex_unlock(&list_lock);
 #endif
 
@@ -986,30 +984,6 @@ get_free_list (void)
 static mstate
 reused_arena (void)
 {
-  if (narenas <= mp_.arena_test)
-    return NULL;
-
-  static int narenas_limit;
-  if (narenas_limit == 0)
-    {
-      if (mp_.arena_max != 0)
-	narenas_limit = mp_.arena_max;
-      else
-	{
-	  int n  = __get_nprocs ();
-
-	  if (n >= 1)
-	    narenas_limit = NARENAS_FROM_NCORES (n);
-	  else
-	    /* We have no information about the system.  Assume two
-	       cores.  */
-	    narenas_limit = NARENAS_FROM_NCORES (2);
-	}
-    }
-
-  if (narenas < narenas_limit)
-    return NULL;
-
   mstate result;
   static mstate next_to_use;
   if (next_to_use == NULL)
@@ -1048,10 +1022,41 @@ arena_get2(a_tsd, size) mstate a_tsd; size_t size;
   mstate a;
 
 #ifdef PER_THREAD
-  if ((a = get_free_list ()) == NULL
-      && (a = reused_arena ()) == NULL)
-    /* Nothing immediately available, so generate a new arena.  */
-    a = _int_new_arena(size);
+  static size_t narenas_limit;
+
+  a = get_free_list ();
+  if (a == NULL)
+    {
+      /* Nothing immediately available, so generate a new arena.  */
+      if (narenas_limit == 0)
+	{
+	  if (mp_.arena_max != 0)
+	    narenas_limit = mp_.arena_max;
+	  else
+	    {
+	      int n  = __get_nprocs ();
+
+	      if (n >= 1)
+		narenas_limit = NARENAS_FROM_NCORES (n);
+	      else
+		/* We have no information about the system.  Assume two
+		   cores.  */
+		narenas_limit = NARENAS_FROM_NCORES (2);
+	    }
+	}
+    repeat:;
+      size_t n = narenas;
+      if (__builtin_expect (n <= mp_.arena_test || n < narenas_limit, 0))
+	{
+	  if (catomic_compare_and_exchange_bool_acq (&narenas, n + 1, n))
+	    goto repeat;
+	  a = _int_new_arena (size);
+	  if (__builtin_expect (a == NULL, 0))
+	    catomic_decrement (&narenas);
+	}
+      else
+	a = reused_arena ();
+    }
 #else
   if(!a_tsd)
     a = a_tsd = &main_arena;
-- 
1.7.8.6

